lib/gc/Transforms/BufferHoist.cpp

```diff
@@ -0,0 +1,433 @@
+  // (currently) not possible to detect a loop.
+  auto regionInterface = dyn_cast<RegionBranchOpInterface>(op);
+  if (!regionInterface)
+    return false;
+
+  return regionInterface.hasLoop();
+}
+/// Return whether the given operation is a loop with sequential execution
+/// semantics.
+static bool isSequentialLoop(Operation *op) {
+  return !op->hasTrait<OpTrait::HasParallelRegion>() && isLoop(op);
+}
+
+/// Return whether the given operation is a loop with parallel execution
+/// semantics.
+static bool isParallelLoop(Operation *op) {
+  return isLoop(op) && op->hasTrait<OpTrait::HasParallelRegion>();
+}
+
+static bool allowHoisting(Operation *op) {
+  auto allocOp = dyn_cast<AllocationOpInterface>(op);
+  return allocOp &&
+         static_cast<uint8_t>(allocOp.getHoistingKind() & HoistingKind::Loop);
+}
+
+struct ForallLoopDepth {
+  llvm::DenseMap<Operation *, int32_t> mapping;
+
+  ForallLoopDepth(Operation *op) {
+    int32_t loopDepth = 0;
+    op->walk<WalkOrder::PreOrder>([&](Operation *op) {
+      if (auto forallOp = dyn_cast<scf::ForallOp>(op)) {
+        if (mapping.find(forallOp) == mapping.end()) {
+          if (forallOp->getParentOp() &&
+              isa<scf::ForallOp>(forallOp->getParentOp()) &&
+              mapping.find(forallOp->getParentOp()) != mapping.end()) {
+            mapping[forallOp] = mapping[forallOp->getParentOp()] + 1;
+          } else {
+            mapping[forallOp] = loopDepth++;
+          }
+        }
+      }
+    });
+  }
+
+  /// Returns the loop depth/level of the given forall op.
+  int32_t getLoopDepth(Operation *op) {
+    auto it = mapping.find(op);
+    if (it != mapping.end())
+      return it->second;
+    return std::numeric_limits<int32_t>::min();
+  }
+};
+
+/// A state implementation compatible with the `BufferAllocationHoisting` class
+/// that hoists allocations out of loops.
+struct BufferHoistingState {
+  /// A pointer to the current dominance info.
+  DominanceInfo *dominators;
+
+  /// The current allocation value.
+  Value allocValue;
+
+  /// The current placement block (if any).
+  Block *placementBlock;
+
+  /// Initializes the state base.
+  BufferHoistingState(DominanceInfo *dominators, Value allocValue,
+                      Block *placementBlock)
+      : dominators(dominators), allocValue(allocValue),
+        placementBlock(placementBlock) {}
+
+  /// Remembers the dominator block of all aliases.
+  Block *aliasDominatorBlock = nullptr;
+
+  /// Computes the upper bound for the placement block search.
+  Block *computeUpperBound(Block *dominatorBlock, Block *dependencyBlock) {
+    aliasDominatorBlock = dominatorBlock;
+    // If there is a dependency block, we have to use this block as an upper
+    // bound to satisfy all allocation value dependencies.
+    return dependencyBlock ? dependencyBlock : nullptr;
+  }
+
+  /// Returns true if the given operation represents a loop with sequential
+  /// execution semantics and one of the aliases caused the
```

Comment:
There is a contradiction between the statement "one of the aliases caused the aliasDominatorBlock to be 'above' the block of the given loop operation" and the condition "!dominators->dominates(aliasDominatorBlock, op->getBlock())"？

Reply:
You're right, the comment here is not correctly reflect the condition below.  Only the `aliasDominatorBlock` is lower (not above) than `op->getBlock()` is the true case.